src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicAnalysisDataReader.cs

```diff
@@ -0,0 +1,367 @@
+        {
+            var reader = new BlobReader(buffer, size);
+
+            // header:
+            if (reader.ReadByte() != 'D' || reader.ReadByte() != 'A' || reader.ReadByte() != 'M' || reader.ReadByte() != 'D')
+            {
+                throw new BadImageFormatException();
+            }
+
+            // version
+            byte major = reader.ReadByte();
+            byte minor = reader.ReadByte();
+            if (major != 0 && minor != 1)
+            {
+                throw new NotSupportedException();
+            }
+
+            // table sizes:
+            int documentRowCount = reader.ReadInt32();
+            int methodSpanRowCount = reader.ReadInt32();
+
+            // blob heap sizes:
+            int stringHeapSize = reader.ReadInt32();
+            int userStringHeapSize = reader.ReadInt32();
+            int guidHeapSize = reader.ReadInt32();
+            int blobHeapSize = reader.ReadInt32();
+
+            // TODO: check size ranges
+
+            bool isBlobHeapSmall = blobHeapSize < 0x10000;
+            bool isGuidHeapSmall = guidHeapSize / GuidSize < 0x10000; // TODO
+
+            var documentsBuilder = ArrayBuilder<DynamicAnalysisDocument>.GetInstance(documentRowCount);
+
+            for (int i = 0; i < documentRowCount; i++)
+            {
+                var name = MetadataTokens.BlobHandle(ReadReference(ref reader, isBlobHeapSmall));
+                var hashAlgorithm = MetadataTokens.GuidHandle(ReadReference(ref reader, isGuidHeapSmall));
+                var hash = MetadataTokens.BlobHandle(ReadReference(ref reader, isBlobHeapSmall));
+
+                documentsBuilder.Add(new DynamicAnalysisDocument(name, hashAlgorithm, hash));
+            }
+
+            Documents = documentsBuilder.ToImmutableAndFree();
+
+            var methodsBuilder = ArrayBuilder<DynamicAnalysisMethod>.GetInstance(methodSpanRowCount);
+
+            for (int i = 0; i < methodSpanRowCount; i++)
+            {
+                methodsBuilder.Add(new DynamicAnalysisMethod(MetadataTokens.BlobHandle(ReadReference(ref reader, isBlobHeapSmall))));
+            }
+
+            Methods = methodsBuilder.ToImmutableAndFree();
+
+            int stringHeapOffset = reader.Offset;
+            int userStringHeapOffset = stringHeapOffset + stringHeapSize;
+            int guidHeapOffset = userStringHeapOffset + userStringHeapSize;
+            int blobHeapOffset = guidHeapOffset + guidHeapSize;
+
+            if (reader.Length != blobHeapOffset + blobHeapSize)
+            {
+                throw new BadImageFormatException();
+            }
+
+            _guidHeapBlob = new Blob(buffer + guidHeapOffset, guidHeapSize);
+            _blobHeapBlob = new Blob(buffer + blobHeapOffset, blobHeapSize);
+        }
+
+        public static DynamicAnalysisDataReader TryCreateFromPE(PEReader peReader)
+        {
+            // TODO: review all range checks, better error messages
```

Comment:
If this won't be included in the initial checkin, can you please file a TestImpact issue for it?